relay-server/src/services/spooler/envelope_stack/sqlite.rs

```diff
@@ -0,0 +1,515 @@
+use crate::envelope::Envelope;
+use crate::extractors::StartTime;
+use crate::services::spooler::envelope_stack::EnvelopeStack;
+use futures::StreamExt;
+use relay_base_schema::project::ProjectKey;
+use sqlx::query::Query;
+use sqlx::sqlite::{SqliteArguments, SqliteRow};
+use sqlx::{Pool, QueryBuilder, Row, Sqlite};
+use std::cmp::Ordering;
+use std::collections::{BinaryHeap, VecDeque};
+use std::error::Error;
+use std::pin::pin;
+
+struct OrderedEnvelope(Box<Envelope>);
+
+impl Eq for OrderedEnvelope {}
+
+impl PartialEq<Self> for OrderedEnvelope {
+    fn eq(&self, other: &Self) -> bool {
+        matches!(self.cmp(other), Ordering::Equal)
+    }
+}
+
+impl PartialOrd<Self> for OrderedEnvelope {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for OrderedEnvelope {
+    fn cmp(&self, other: &Self) -> Ordering {
+        received_at(&other.0).cmp(&received_at(&self.0))
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SQLiteEnvelopeStackError {
+    #[error("the stack is empty")]
+    Empty,
+
+    #[error("a database error occurred")]
+    DatabaseError(#[from] sqlx::Error),
+}
+
+pub struct SQLiteEnvelopeStack {
+    db: Pool<Sqlite>,
+    spool_threshold: usize,
+    own_key: ProjectKey,
+    sampling_key: ProjectKey,
+    buffer: VecDeque<Box<Envelope>>,
+}
+
+impl SQLiteEnvelopeStack {
+    #[allow(dead_code)]
+    pub fn new(
+        db: Pool<Sqlite>,
+        spool_threshold: usize,
+        own_key: ProjectKey,
+        sampling_key: ProjectKey,
+    ) -> Self {
+        Self {
+            db,
+            spool_threshold,
+            own_key,
+            sampling_key,
+            buffer: VecDeque::with_capacity(spool_threshold),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn prepare(
+        db: Pool<Sqlite>,
+        spool_threshold: usize,
+        own_key: ProjectKey,
+        sampling_key: ProjectKey,
+    ) -> Result<(), SQLiteEnvelopeStackError> {
+        let mut stack = Self::new(db, spool_threshold, own_key, sampling_key);
+        stack.load_from_disk().await?;
+
+        Ok(())
+    }
+
+    fn above_spool_threshold(&self) -> bool {
+        self.buffer.len() + 1 > self.spool_threshold
+    }
+
+    fn below_unspool_threshold(&self) -> bool {
+        self.buffer.is_empty()
+    }
+
+    fn disk_batch_size(&self) -> usize {
+        self.spool_threshold / 2
+    }
+
+    async fn spool_to_disk(&mut self) -> Result<(), SQLiteEnvelopeStackError> {
+        if self.disk_batch_size() == 0 {
+            return Ok(());
+        }
+
+        // TODO: we can make a custom iterator to consume back elements until threshold to avoid
+        //  allocating a vector.
+        let mut envelopes = Vec::with_capacity(self.disk_batch_size());
```

Comment:
I thought about making a custom iterator to support this use case, might be a good alternative to allocating an array.

Reply:
It might be convenient to change `self.buffer` to a `VecDeque<Vec<_>>`, i.e. a deque of batches. Then when we spool or unspool, we just push or pop a single element of the deque.

Reply:
What if we use a type that can be either a single value of a vector? I would like to avoid allocating a vector with 1 element every time I do a push.

Reply:
For pushing, I was thinking of something like

```rust
if self.batches.last().is_none_or(|last_batch| last_batch.len() >= batch_size) {
    self.batches.push(Vec::with_capacity(batch_size))
}
self.batches.last_mut().unwrap().push(envelope);
```

Reply:
Ahh so you would like to store individual batches, it seems a good idea!